File: lang/python/cxlib/cxpath.py
"""Tree-mutation helpers used by Document.transform / transform_all.

The CXPath parser/lexer/evaluator/predicate types that previously lived
here were deleted in Phase 4 (CB-5): bindings now thunk to libcx for
expression evaluation via `cx.select_all_paths`. Only the host-language
tree-mutation utilities remain — those are not duplication, they are
Python-level structural editing.

The utilities here are:
  - elem_detached(e):       deep-ish copy so user callbacks can't mutate source.
  - doc_replace_at(d, i):   functional replace of doc.elements[i].
  - elem_replace_item_at:   functional replace of element.items[i].
  - path_copy_element:      Document.transform helper (slash-path semantics).
  - replace_at_doc_path:    path-based replacement used by transform_all.
"""
from __future__ import annotations
from typing import Any, Callable, Optional, Sequence


def elem_detached(e) -> Any:
    """Return a copy of e with independent attrs/items lists so f cannot mutate the source."""
    from .ast import Element, Attr
    return Element(
        name=e.name,
        anchor=e.anchor,
        merge=e.merge,
        data_type=e.data_type,
        attrs=[Attr(a.name, a.value, a.data_type) for a in e.attrs],
        items=list(e.items),
    )


def doc_replace_at(d, idx: int, el) -> Any:
    from .ast import Document
    return Document(
        elements=[el if i == idx else n for i, n in enumerate(d.elements)],
        prolog=d.prolog,
        doctype=d.doctype,
    )


def elem_replace_item_at(e, idx: int, child) -> Any:
    from .ast import Element
    return Element(
        name=e.name,
        anchor=e.anchor,
        merge=e.merge,
        data_type=e.data_type,
        attrs=e.attrs,
        items=[child if i == idx else n for i, n in enumerate(e.items)],
    )


def path_copy_element(e, parts: list, f: Callable) -> Optional[Any]:
    """Returns a new Element with f applied at parts[...], or None if path not found."""
    from .ast import Element
    for i, item in enumerate(e.items):
        if isinstance(item, Element) and item.name == parts[0]:
            if len(parts) == 1:
                return elem_replace_item_at(e, i, f(elem_detached(item)))
            updated = path_copy_element(item, parts[1:], f)
            if updated is not None:
                return elem_replace_item_at(e, i, updated)
            return None
    return None


# ── Path-based replacement (CB-5 / Phase 4) ──────────────────────────────────

def navigate_doc_path(doc, path: Sequence[int]):
    """Walk `path` (indices into Document.elements then Element.items)
    and return the element at that position. Returns None if any step is
    out-of-bounds or hits a non-Element item.
    """
    from .ast import Element
    if not path:
        return None
    if path[0] >= len(doc.elements):
        return None
    node = doc.elements[path[0]]
    for k in path[1:]:
        if not isinstance(node, Element):
            return None
        if k >= len(node.items):
            return None
        node = node.items[k]
    return node if isinstance(node, Element) else None


def replace_at_doc_path(doc, path: Sequence[int], new_elem) -> Any:
    """Return a new Document with the element at `path` replaced by
    `new_elem`. Original document is unchanged (functional update —
    only the spine along `path` is rebuilt).
    """
    from .ast import Document
    if not path:
        raise ValueError("replace_at_doc_path: empty path")
    new_elements = list(doc.elements)
    if len(path) == 1:
        new_elements[path[0]] = new_elem
    else:
        top = new_elements[path[0]]
        new_elements[path[0]] = _replace_in_element(top, path[1:], new_elem)
    return Document(
        elements=new_elements,
        prolog=doc.prolog,
        doctype=doc.doctype,
    )


def _replace_in_element(el, path: Sequence[int], new_elem):
    from .ast import Element
    if not path:
        return new_elem
    new_items = list(el.items)
    if len(path) == 1:
        new_items[path[0]] = new_elem
    else:
        child = new_items[path[0]]
        new_items[path[0]] = _replace_in_element(child, path[1:], new_elem)
    return Element(
        name=el.name,
        anchor=el.anchor,
        merge=el.merge,
        data_type=el.data_type,
        attrs=list(el.attrs),
        items=new_items,
    )

File: lang/csharp/cxlib/CXPath.cs
using System;
using System.Collections.Generic;

namespace CX;

/// <summary>
/// Tree-mutation helpers used by <see cref="CXDocument.Transform"/> /
/// <see cref="CXDocument.TransformAll"/>.
///
/// <para>The CXPath parser, lexer, predicate types, and evaluator that
/// lived here were deleted in Phase 4 (CB-5): expression evaluation
/// now runs in libcx via <c>cx_select_all_paths</c>, and bindings
/// navigate the returned structural paths against their in-memory tree.
/// What remains here is host-language tree manipulation — not
/// duplication.</para>
/// </summary>
public static class CXPath
{
    /// <summary>Return a copy of <paramref name="e"/> with independent
    /// attrs/items lists. Attr is a record (value-equal, immutable);
    /// the shallow list copy is safe because SetAttr replaces entries,
    /// it doesn't mutate them.</summary>
    public static Element ElemDetached(Element e)
    {
        var copy = new Element(e.Name)
        {
            Anchor = e.Anchor,
            Merge = e.Merge,
            DataType = e.DataType,
        };
        copy.Attrs = new List<Attr>(e.Attrs);
        copy.Items = new List<Node>(e.Items);
        return copy;
    }

    /// <summary>Functional replace of d.Elements[idx].</summary>
    public static CXDocument DocReplaceAt(CXDocument d, int idx, Element el)
    {
        var newElements = new List<Node>(d.Elements);
        if (idx >= 0 && idx < newElements.Count) newElements[idx] = el;
        return new CXDocument { Elements = newElements, Prolog = d.Prolog, Doctype = d.Doctype };
    }

    /// <summary>Functional replace of e.Items[idx].</summary>
    public static Element ElemReplaceItemAt(Element e, int idx, Node child)
    {
        var newItems = new List<Node>(e.Items);
        if (idx >= 0 && idx < newItems.Count) newItems[idx] = child;
        return new Element(e.Name)
        {
            Anchor = e.Anchor,
            Merge = e.Merge,
            DataType = e.DataType,
            Attrs = e.Attrs,
            Items = newItems,
        };
    }

    /// <summary>Slash-path-style descent for <see cref="CXDocument.Transform"/>.</summary>
    public static Element? PathCopyElement(Element e, string[] parts, Func<Element, Element> f)
    {
        for (int i = 0; i < e.Items.Count; i++)
        {
            if (e.Items[i] is Element item && item.Name == parts[0])
            {
                if (parts.Length == 1)
                    return ElemReplaceItemAt(e, i, f(ElemDetached(item)));
                var updated = PathCopyElement(item, parts[1..], f);
                if (updated is not null)
                    return ElemReplaceItemAt(e, i, updated);
                return null;
            }
        }
        return null;
    }

    // ── Path-based navigation (CB-5 / Phase 4) ────────────────────────────────

    /// <summary>
    /// Walk <paramref name="path"/> (indices into Document.Elements then
    /// Element.Items) and return the live element reference at that
    /// position. Returns null if any step is out-of-bounds or hits a
    /// non-Element item.
    /// </summary>
    public static Element? NavigateDocPath(CXDocument d, int[] path)
    {
        if (path.Length == 0 || path[0] < 0 || path[0] >= d.Elements.Count) return null;
        Node node = d.Elements[path[0]];
        for (int i = 1; i < path.Length; i++)
        {
            if (node is not Element el) return null;
            int k = path[i];
            if (k < 0 || k >= el.Items.Count) return null;
            node = el.Items[k];
        }
        return node as Element;
    }

    /// <summary>
    /// Return a new <see cref="CXDocument"/> with the element at
    /// <paramref name="path"/> replaced by <paramref name="newElem"/>.
    /// Only the spine along <paramref name="path"/> is rebuilt; the
    /// original document is unchanged.
    /// </summary>
    public static CXDocument ReplaceAtDocPath(CXDocument d, int[] path, Element newElem)
    {
        if (path.Length == 0) return d;
        var newElements = new List<Node>(d.Elements);
        if (path.Length == 1)
        {
            if (path[0] >= 0 && path[0] < newElements.Count)
                newElements[path[0]] = newElem;
        }
        else if (newElements[path[0]] is Element top)
        {
            var rest = path[1..];
            newElements[path[0]] = ReplaceInElement(top, rest, newElem);
        }
        return new CXDocument { Elements = newElements, Prolog = d.Prolog, Doctype = d.Doctype };
    }

    private static Element ReplaceInElement(Element el, int[] path, Element newElem)
    {
        if (path.Length == 0) return newElem;
        var newItems = new List<Node>(el.Items);
        if (path.Length == 1)
        {
            if (path[0] >= 0 && path[0] < newItems.Count) newItems[path[0]] = newElem;
        }
        else if (newItems[path[0]] is Element child)
        {
            var rest = path[1..];
            newItems[path[0]] = ReplaceInElement(child, rest, newElem);
        }
        return new Element(el.Name)
        {
            Anchor = el.Anchor,
            Merge = el.Merge,
            DataType = el.DataType,
            Attrs = el.Attrs,
            Items = newItems,
        };
    }
}
